from google.coab import files
uploaded=files.upload()[link text](https://)

In [5]:
from google.colab import files
uploaded=files.upload()

Saving adult.csv to adult.csv


In [19]:
#task1
import pandas as pd
import numpy as np

df = pd.read_csv("adult.csv")

df.head()
df.info()

# missing values
print("Missing values before cleaning:")
print(df.isnull().sum())

df.replace("?", np.nan, inplace=True)

print("\nMissing values after cleaning:")
print(df.isnull().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB
Missing values before cleaning:
age               0
workclass         0
fnlwgt            0
e

In [20]:
#task2
missing_sum = df.isnull().sum()
missing_percentage = df.isnull().mean() * 100

summary = pd.DataFrame({
    "Missing Count": missing_sum,
    "Missing Percentage": missing_percentage
})
summary = summary.sort_values(
    by="Missing Percentage",
    ascending=False
)
summary

,Missing Count,Missing Percentage
occupation,1843,5.660146
workclass,1836,5.638647
native.country,583,1.790486
fnlwgt,0,0.000000
education,0,0.000000
education.num,0,0.000000
age,0,0.000000
marital.status,0,0.000000
relationship,0,0.000000
sex,0,0.000000


In [21]:
#task3

df["workclass"] = df["workclass"].fillna("Unknown")

df["occupation"] = df["occupation"].fillna("Unknown")

df["native.country"] = df["native.country"].fillna("Unknown")

df.isnull().sum()

,0
age,0
workclass,0
fnlwgt,0
education,0
education.num,0
marital.status,0
occupation,0
relationship,0
race,0
sex,0


In [ ]:
task3 reasoning:
For workclass and occupation, I decided to replace the missing values with "Unknown" instead of deleting the rows. Since these are categorical columns, using "Unknown" keeps the data and does not assume any specific category.

I also replaced the missing values in native.country with "Unknown" for the same reason. There were no missing values in any numerical columns, so no numerical imputation was needed.

In [23]:
#TASK4
print("Exact duplicate rows:", df.duplicated().sum())

features = df.drop(columns=['income'])

print("Duplicates ignoring target:", features.duplicated().sum())

print("Decision: Rows with identical attributes but different income outcomes are not treated as duplicates.")

Exact duplicate rows: 24
Duplicates ignoring target: 25
Decision: Rows with identical attributes but different income outcomes are not treated as duplicates.


In [27]:
#task5


for col in ['education', 'marital.status', 'native.country']:
    print("\nUnique values in", col, "before cleaning:")
    print(df[col].unique())

for col in ['education', 'marital.status', 'native.country']:
    df[col] = df[col].str.strip()

for col in ['education', 'marital.status', 'native.country']:
    print("\nUnique values in", col, "after cleaning:")
    print(df[col].unique())


Unique values in education before cleaning:
['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']

Unique values in marital.status before cleaning:
['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']

Unique values in native.country before cleaning:
['United-States' 'Unknown' 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan'
 'India' 'Philippines' 'Trinadad&Tobago' 'Canada' 'South'
 'Holand-Netherlands' 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany'
 'Italy' 'Japan' 'Hong' 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru'
 'Nicaragua' 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary'
 'Columbia' 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia'
 'Scotland' 'Portugal' 'Laos' 'Thailand' 'Outlying-US(Guam-USVI-etc)']

Unique values in education after cleaning:
['HS-grad' 'Some-college' '7th-8th' '10th' 'Doct

In [28]:
#task6

print("Numerical Descriptive Statistics:")
print(df.describe())

print("\nCategorical Descriptive Statistics:")
print(df.describe(include='object'))

print("\nMean and Median:")
print("Age Mean:", df['age'].mean())
print("Age Median:", df['age'].median())
print("Hours-per-week Mean:", df['hours.per.week'].mean())
print("Hours-per-week Median:", df['hours.per.week'].median())

print("\nComment:")
print("The mean and median are compared to determine whether the distributions are noticeably skewed.")

Numerical Descriptive Statistics:
                age        fnlwgt  education.num  capital.gain  capital.loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours.per.week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000  

Categorical Descriptive Stat

In [29]:
#task 7

print("Most Common Occupation:")
print(df['occupation'].value_counts().head(1))

print("\nSex Percentage Distribution:")
print((df['sex'].value_counts(normalize=True) * 100).round(2))

print("\nIncome Percentage Distribution:")
print((df['income'].value_counts(normalize=True) * 100).round(2))

print("\nIncome Balance:")
income_percent = df['income'].value_counts(normalize=True) * 100

if abs(income_percent.iloc[0] - income_percent.iloc[1]) < 10:
    print("The income classes are relatively balanced.")
else:
    print("The income classes are imbalanced.")

Most Common Occupation:
occupation
Prof-specialty    4140
Name: count, dtype: int64

Sex Percentage Distribution:
sex
Male      66.92
Female    33.08
Name: proportion, dtype: float64

Income Percentage Distribution:
income
<=50K    75.92
>50K     24.08
Name: proportion, dtype: float64

Income Balance:
The income classes are imbalanced.


In [32]:
#task8

education_check = df.groupby('education')['education.num'].unique()

print("Education and education.num mapping:")
print(education_check)

inconsistencies = education_check[education_check.apply(len) > 1]

print("\nInconsistencies found:")
if len(inconsistencies) == 0:
    print("No inconsistencies found.")
else:
    print(inconsistencies)

Education and education.num mapping:
education
10th             [6]
11th             [7]
12th             [8]
1st-4th          [2]
5th-6th          [3]
7th-8th          [4]
9th              [5]
Assoc-acdm      [12]
Assoc-voc       [11]
Bachelors       [13]
Doctorate       [16]
HS-grad          [9]
Masters         [14]
Preschool        [1]
Prof-school     [15]
Some-college    [10]
Name: education.num, dtype: object

Inconsistencies found:
No inconsistencies found.


In [34]:
#task 9
import pandas as pd
import sqlite3

df_original = pd.read_csv("adult.csv")
conn = sqlite3.connect("adult_income.db")

df_original.to_sql("adult_income", conn, if_exists="replace", index=False)
query = "SELECT * FROM adult_income WHERE age > 30"

df_sql = pd.read_sql_query(query, conn)

print("Original CSV shape:", df_original.shape)
print("SQL result shape:", df_sql.shape)

print("\nSQL extracted data:")
print(df_sql.head())

conn.close()

Original CSV shape: (32561, 15)
SQL result shape: (21989, 15)

SQL extracted data:
   age workclass  fnlwgt     education  education.num marital.status  \
0   90         ?   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66         ?  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship   race     sex  capital.gain  \
0                  ?  Not-in-family  White  Female             0   
1    Exec-managerial  Not-in-family  White  Female             0   
2                  ?      Unmarried  Black  Female             0   
3  Machine-op-inspct      Unmarried  White  Female             0   
4     Prof-specialty      Own-child  White  Female             0   

   capital.loss  hours.per.week native.country income  
0          4356    

# Task 10
## 1. Dataset Overview

The dataset used is the UCI Adult Income Dataset. It contains information about individuals from the 1994 US Census and is used to predict whether a person's annual income is greater than $50K.

The original dataset contains **32,561 rows and 15 columns**. It includes **6 numerical columns and 9 categorical columns**.

## 2. Data Quality Issues Found

The dataset contained missing values represented by the `?` symbol:

- `occupation`: 1,843 missing values
- `workclass`: 1,836 missing values
- `native.country`: 583 missing values

There were **24 exact duplicate rows** in the original dataset.

There were also leading/trailing spaces in categorical values, which could create inconsistent categories.

The `education` and `education.num` columns were checked and found to be consistent, with each education level mapping to the same education number.

## 3. Key Observations

The most common occupation was **Prof-specialty**, with **4,136** records.

The sex distribution was:

- Male: **66.92%**
- Female: **33.08%**

The income distribution was:

- `<=50K`: **75.91%**
- `>50K`: **24.09%**

Therefore, the income classes are **imbalanced**, because most individuals belong to the `<=50K` class.

The mean hours-per-week was **40.44 hours**, while the median was **40 hours**. These values are very close, suggesting that the typical working week is around 40 hours.

## 4. Cleaning Decisions Made

The `?` values were treated as missing values and replaced with `NaN`. The missing categorical values were then filled with `Unknown` instead of removing those rows, so that useful observations were not lost.

The **24 exact duplicate rows** were removed.

Leading and trailing whitespace was removed from categorical columns using `.str.strip()`.

No columns were removed. The `education` and `education.num` relationship was checked and found to be consistent.